### PR TITLE
fix(client): Randomly slow youtube download speed

### PR DESCRIPTION
**Pull request by jnelle**
# Description

This PR fixes the randomly slow youtube download speed that caused from web client.

## Issues to fix

Please link issues this PR will fix: \
[\#_[230]_](https://github.com/kkdai/youtube/issues/230)

## Reminding
Something you can do before PR to reduce time to merge

* run "make build" to build the code
* run "make format" to reformat the code
* run "make lint" if you are using unix system
* run "make test-integration" to pass all tests

### CODE REVIEW

**Comment by corny** [comment]
Thanks a lot for you contribution! Unfortunately one test fails because the changed user agent leads to a missing creation timestamp. I could not find a solution yet.

**Comment by jnelle** [comment]
> Thanks a lot for you contribution! Unfortunately one test fails because the changed user agent leads to a missing creation timestamp. I could not find a solution yet.

Thanks for your help! I'll have a closer look this evening

File: video.go [hunk 1]
```diff
@@ -95,18 +95,19 @@ func (v *Video) extractDataFromPlayerResponse(prData playerResponseData) error {
...
 
-	if seconds, _ := strconv.Atoi(prData.Microformat.PlayerMicroformatRenderer.LengthSeconds); seconds > 0 {
+	if seconds, _ := strconv.Atoi(prData.VideoDetails.LengthSeconds); seconds > 0 {
 		v.Duration = time.Duration(seconds) * time.Second
 	}
 
-	if str := prData.Microformat.PlayerMicroformatRenderer.PublishDate; str != "" {
-		v.PublishDate, _ = time.Parse(dateFormat, str)
-	}
+	// couldn't find any publishdate for now, have to add an another request
```
Comment (jnelle) [inline video.go:104]: @corny could you check this please?

File: video.go [hunk 2]
```diff
@@ -95,18 +95,19 @@ func (v *Video) extractDataFromPlayerResponse(prData playerResponseData) error {
...
+	// if str := prData.VideoDetails.Microformat.PlayerMicroformatRenderer.PublishDate; str != "" {
+	// 	v.PublishDate, _ = time.Parse(dateFormat, str)
+	// }
 
 	// Assign Streams
-	v.Formats = append(prData.StreamingData.Formats, prData.StreamingData.AdaptiveFormats...)
+	v.Formats = append(prData.StreamingData.Formats, prData.StreamingData.Formats...)
```
Comment (cybre) [inline video.go:110]: Why are you duplicating the `StreamingData.Formats` slice?

**Comment by cybre** [comment]
I ran into the issue with slow downloads and tried this patch out and noticed some issues.
This PR introduces a new type for adaptive formats: `AdaptiveFormat`. The newly separated adaptive formats are thus not included in the `Video` type so we completely lose access to them. Additionally, even if they were exposed, `Client` in its current state only works with the `Format` type. Would it be possible switch back to a single `Format` type?

**Comment by jnelle** [comment]
> I ran into the issue with slow downloads and tried this patch out and noticed some issues. This PR introduces a new type for adaptive formats: `AdaptiveFormat`. The newly separated adaptive formats are thus not included in the `Video` type so we completely lose access to them. Additionally, even if they were exposed, `Client` in its current state only works with the `Format` type. Would it be possible switch back to a single `Format` type or create a format interface?

https://github.com/cybre/youtube/commit/4a0c7d672485d96144fa26aebbfc602b3ad2bb2f

I think your solution in this commit is the best for now!

File: video.go (hunk 2 above)
Comment (jnelle) [reply video.go:110]: https://github.com/cybre/youtube/commit/4a0c7d672485d96144fa26aebbfc602b3ad2bb2f

I forgot to change it from AdaptiveFormats to Format like you did in your solution

**Comment by jnelle** [comment]
@corny @kkdai, should I fix the conflict and we merge this change?

**Comment by corny** [comment]
Please rebase on top of master and then let's check the integration tests.

**Comment by corny** [comment]
I tried to rebase your branch on top of master. Unfortuntely test `TestYoutube_GetItagInfo` shows that the number of itags drops from 26 to 6 for one video and there are more issues.

**Comment by 89z** [comment]
FYI I am the author of the original `"clientVersion": "16.20"` code:

https://github.com/89z/mech/commit/f1b0efc090fc2e72e5950380a4ff54f6d92a22ec

I would recommend changing that, as at the time I didnt know how the version was derived, so I created it from trial and error. Ive since learned that its just the version number from the Android app:

~~~
PS C:\> googleplay -a com.google.android.youtube
POST https://android.googleapis.com/auth
GET https://android.clients.google.com/fdfe/details?doc=com.google.android.youtube
Title: YouTube
Creator: Google LLC
UploadDate: Mar 16, 2022
VersionString: 17.11.34
~~~

**Comment by corny** [comment]
should be solved by #244

**Pull request closed without merging**